diff --git a/config.def.h b/config.def.h
index ef44721..8d95713 100644
--- a/config.def.h
+++ b/config.def.h
@@ -34,7 +34,7 @@ static Parameter defconfig[ParameterLast] = {
 	[JavaScript]          =       { { .i = 1 },     },
 	[KioskMode]           =       { { .i = 0 },     },
 	[LoadImages]          =       { { .i = 1 },     },
-	[MediaManualPlay]     =       { { .i = 1 },     },
+	[MediaManualPlay]     =       { { .i = 0 },     },
 	[PreferredLanguages]  =       { { .v = (char *[]){ NULL } }, },
 	[RunInFullscreen]     =       { { .i = 0 },     },
 	[ScrollBars]          =       { { .i = 1 },     },
